src/cfnlint/data/AdditionalSpecs/SubNeededExcludes.json

```diff
@@ -56,89 +29,24 @@
                 "AWS::CloudFormation::Init": {
                     "ExcludeAll": true
                 }
-            },
-            "Properties": {
-                "UserData": {
-                    "ExcludeAll": true
-                }
             }
         },
         "AWS::EMR::InstanceGroupConfig": {
             "Properties": {
                 "AutoScalingPolicy": {
                     "Properties": {
-                        "ScalingRule": {
+                        "Rules": {
                             "Properties": {
-                                "ScalingTrigger": {
+                                "Trigger": {
                                     "Properties": {
-                                       "CloudWatchAlarmDefinition": {
+                                        "CloudWatchAlarmDefinition": {
                                             "ExcludeAll": true
```

Comment:
[`AWS::EMR::Cluster.ScalingTrigger.CloudWatchAlarmDefinition`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticmapreduce-cluster-scalingtrigger.html#cfn-elasticmapreduce-cluster-scalingtrigger-cloudwatchalarmdefinition)

https://github.com/aws-cloudformation/cfn-python-lint/pull/1788#issuecomment-731345937 was just meant to be a few examples rather than an exhaustive list

I'm not sure I understand the benefit over the previous format:

https://github.com/aws-cloudformation/cfn-python-lint/blob/52c02962e90354b265fec839e71e1d038e899fd4/src/cfnlint/rules/functions/SubNeeded.py#L21-L23

since it's now hundreds of line instead with decreased coverage and increased maintenance

Reply:
Prior to #1788, there was a lot of custom logic to handle things such as IAM and IoT policies, where there's a specific list of permitted substitutions the rule should not match, and API gateway substitutions, where there's a regex to match permitted substitutions. I decided to make the rule more tolerant to such use-cases, and that moving the exclusions into a separate configuration file was a sensible addition. Should the resource provider schema ever include such information itself, then the rule is in a much better position to take advantage of it.

I fully recognise the specs file is a lot bigger than the original exclusion variables, however I think it's also a lot more descriptive, and I disagree around the decreased coverage and increased maintenance. With this PR, the specs file allows you to exclude a property regardless of what resource or path within a resource it is (using the `GlobalPropertyTypes` key), so I just need to move the `CloudWatchAlarmDefinition` block to there to cover both the `AWS::EMR::Cluster` and `AWS::EMR::InstanceGroupConfig`.

Maybe it makes more sense for everything to be under `GlobalPropertyTypes` rather than `ResourceTypes`, however I was keen to avoid false-negatives if two resources have a property with the same name, but with different behaviours. Moving all the current blocks to under `GlobalPropertyTypes` will emulate the original `excludes` variable's behaviour.

Reply:
> Should the resource provider schema ever include such information itself

This is unlikely btw

> With this PR, the specs file allows you to exclude a property regardless of what resource or path within a resource it is

believe this was already the case

> however I was keen to avoid false-negatives if two resources have a property with the same name, but with different behaviours

Agree linters should strive for false positives over false negatives, but I think this is actually more likely to cause false negatives than simply excluding all properties with the same name

Reply:
same issue with [`AWS::CodeBuild::Project.Source.BuildSpec`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-codebuild-project-source.html#cfn-codebuild-project-source-buildspec) not working anymore